convert2rhel/checks.py

```diff
@@ -287,66 +296,83 @@ def get_most_recent_unique_kernel_pkgs(pkgs):
+
+    For example, consider the following list of packages::
+
+        list_of_pkgs = [
+            'kernel-core-0:4.18.0-240.10.1.el8_3.x86_64',
+            'kernel-core-0:4.19.0-240.10.1.el8_3.x86_64',
+            'kmod-debug-core-0:4.18.0-240.10.1.el8_3.x86_64',
+            'kmod-debug-core-0:4.18.0-245.10.1.el8_3.x86_64
         ]
+
+    And when this function gets called with that same list of packages,
+    we have the following output::
+
+        result = get_most_recent_unique_kernel_pkgs(pkgs=list_of_pkgs)
+        print(result)
+        # (
+        #   'kernel-core-0:4.19.0-240.10.1.el8_3.x86_64',
+        #   'kmod-debug-core-0:4.18.0-245.10.1.el8_3.x86_64'
+        # )
+
+    :param pkgs: A list of package names to be analyzed.
+    :type pkgs: list[str]
+    :return: A tuple of packages name sorted and normalized
+    :rtype: tuple[str]
     """
 
     pkgs_groups = itertools.groupby(sorted(pkgs), lambda pkg_name: pkg_name.split(":")[0])
-    return tuple(
-        max(distinct_kernel_pkgs[1], key=_repos_version_key)
-        for distinct_kernel_pkgs in pkgs_groups
-        if distinct_kernel_pkgs[0].startswith(("kernel", "kmod"))
-    )
+    list_of_sorted_pkgs = []
+    for distinct_kernel_pkgs in pkgs_groups:
+        if distinct_kernel_pkgs[0].startswith(("kernel", "kmod")):
+            list_of_sorted_pkgs.append(max(distinct_kernel_pkgs[1], key=_repos_version_key))
+
+    return tuple(list_of_sorted_pkgs)
 
 
 def _repos_version_key(pkg_name):
+    """Identify the version key in a given package name.
+
+    Consider the following pkg_name that will be passed to this function::
+
+        pkg_name = 'kernel-core-0:4.18.0-240.10.1.el8_3.x86_64'
+
+    The output of this will be a tuple containing the package version in a tuple::
+
+        result = _repos_version_key(pkg_name=pkg_name)
+        print(result)
+        # (4, 15, 0, 240, 10, 1)
+
+    The function will ignore the package name as it is not an important information here
+    and will only care about the version that is tied to it's name.
+
+    :param pkg_name: The package to extract the version
+    :type pkg_name: str
+    :return: A tuple containing the package version.
+    :rtype: tuple[int]
+    :raises:
+        SystemExit: Raises SytemExit if it can't find the version in the pkg_name.
+    """
+    # TODO(r0x0d): This should be moved to repo.py or utils.py?
     try:
         rpm_version = KERNEL_REPO_RE.search(pkg_name).group("version")
     except AttributeError:
         logger.critical(
-            "Unexpected package:\n%s\n is a source of kernel modules.",
+            "Unexpected package: %s\n Couldn't find the version of the given package.",
             pkg_name,
         )
     else:
         return tuple(
             map(
-                _convert_to_int_or_zero,
+                convert_to_int_or_zero,
                 KERNEL_REPO_VER_SPLIT_RE.split(rpm_version),
             )
         )
```

Comment:
I was unsure about this... Maybe we should move this function to `convert2rhel/repos.py` file? Since it's dealing with "repos" even it is only extracting the version from a string.

Not sure if the `convert2rhel/repos.py` or `convert2rhel/utils.py` is a better location for this function.

Reply:
It feels like maybe a function for pkghandler?  (Unless we want pkghandler to only deal with packages-in-repositories).  It's parsing a version from a single package's NEVR.  One possible reason not to move this (or to change it when we do move it), though: right now, this function is very specifically targeted at the NEVR of kernel package's that we think we are going to encounter.  All of the following NEVRs would output the same value from this function despite being different packages:

* kernel-2.4.0-1.el8_1
* kernel-2.4-0.1.el8_1
* kernel-2.4.0-1.el9_0